OneSignal/OneSignal-iOS-SDK
# [user model] adding the operations layer

# Description
## One Line Summary
Add the base layer for the operation repo and set up some functionality on top of it.

## Details
* Sorry, commit history is not nice.

### Motivation
We need the to implement the operations for the user model.

### Details
**The `OSOperationRepo`:**
- A static singleton class that maintains a queue of OSDeltas, which are divvy'd up at intervals to concrete executors.
- We cache and load operations and deltas in the `OSOperationRepo` and the concrete executors

**Caching deltas and operations:**

1. When `OSOperationRepo` is initialized, read its deltas from cache, if any.
2. When executors are started, read its deltas and operations from cache, if any.
3. When a delta is enqueued into the `OSOperationRepo`, we snapshot the entire deltaQueue and cache.
4. When the operation repo is flushed, we snapshot the deltas remaining in the operation repo, and snapshot the deltas in the executors.
5. When concrete executors' deltas are processed into operations and enqueued, the entire operationQueue is cached after adding. The deltas are snapshotted as well. The details of this isn't fleshed out and will depend on the implementation.
6. When an operation executes, remove it from the operation queue, but not from cache until it returns successfully.
7. When operation returns successfully, remove this operation from cache (and remove from queue if it still exists). This can happen if the app restarts before an operation returns successfully - we will have loaded all the incomplete operations into the operation queue, including the one that just returned successfully.

# Testing
Unit testing done is just for checking method and property access, and relying on printed statements and debug stop points to see what is happening.

* use existing test that is renamed `testModelAndOperationRepositoryHookUpWithLoginAndSetAlias`
* call various adding and removing aliases methods for testing the deltas/operations logic and hookup

## Manual testing
Not implemented enough to test manually properly.

# Affected code checklist
   - [x] User Model changes

# Checklist
## Overview

## Testing

## Final pass


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1118)
